python/ray/util/sgd/data/dataset.py

```diff
@@ -0,0 +1,92 @@
+from ray.util.iter import ParallelIterator, from_iterators
+
+
+class Dataset():
+    """A simple Dataset abstraction for RaySGD.
+
+    This dataset is designed to work with RaySGD trainers (currently just
+    Torch) to provide support for streaming large external datasets, and built
+    in sharding.
+
+    .. code-block:: python
+
+            def to_mat(x):
+                return torch.tensor([[x]]).float()
+
+
+            data = [i * 0.001 for i in range(1000)]
+            p_iter = iter.from_items(data, num_shards=1, repeat=True)
+            dataset = Dataset(
+                p_iter,
+                batch_size=32,
+                max_concurrency=1,
+                download_func=lambda x: (to_mat(x), to_mat(x)))
+
+            trainer = TorchTrainer(
+                model_creator=model_creator,
+                data_creator=None,
+                optimizer_creator=optimizer_creator,
+                loss_creator=torch.nn.MSELoss,
+                num_workers=5,
+            )
+
+            for i in range(10):
+                # Train for another epoch using the dataset
+                trainer.train(dataset=dataset, num_steps=200)
+
+            model = trainer.get_model()
+            print("f(0.5)=", float(model(to_mat(0.5))[0][0]))
+
+    Args:
+        data (iterable[U] or ParallelIterator[U]): Any existing python
+            iterable (or iterator), or an existing parallel iterator
+            to use.
+        batch_size (int): The batch size for training/inference (default 32).
+        download_func (U -> (S, Y)): A function which returns two values, the
+            input and the label (default is the identity function).
+        max_concurrency (int): The maximum number of concurrent calls to the
+            download function. See ParallelIterator::for_each for details.
+        transform (S -> X): A final transformation to be applied to the _input
+            only_. This is guaranteed to run on the same worker that training
+            will occur on.
+    """
+
+    def __init__(self,
+                 data,
+                 batch_size=32,
+                 download_func=None,
+                 max_concurrency=0,
+                 transform=None):
+        par_iter = None
+        if isinstance(data, ParallelIterator):
+            par_iter = data.repartition(1)
```

Comment:
Is there a reason we have to repartition to 1 here instead of leaving as is?

Reply:
Original rationale was so that max_concurrency would set the cluster-wide parallelism. Do you think we should just take it out?

Reply:
I think the original design with forcing exactly `max_concurrency` shards makes more sense. Is there a compelling reason (maybe performance-wise or where in the cluster shards are placed) to use pre-sharded iterators?

Reply:
There are some use cases where this could lead to a lot of data shuffling, but it wouldn't be a very good use of parallel iterator to begin with. 

I propose we leave the repartition here, then remove/rethink the signature if there's a compelling performance reason.

Reply:
OK that sounds good